src/server/editorServices.ts

```diff
@@ -754,6 +756,103 @@ namespace ts.server {
             return false;
         }
 
+        acquireTypingForJs(project: Project) {
+            if (!this.host.getTsd) {
+                return;
+            }
+            const tsd = this.host.getTsd();
+            if (project === undefined || tsd === undefined) {
+                return;
+            }
+
+            const typingOptions: TypingOptions = project.projectOptions
+                ? project.projectOptions.typingOptions
+                : { enableAutoDiscovery: true };
+            const compilerOptions: CompilerOptions = project.projectOptions
+                ? project.projectOptions.compilerOptions
+                : undefined;
+            const projectRootPath: Path = project.isConfiguredProject() ? getDirectoryPath(<Path>project.projectFilename) : undefined;
+
+            // Todo: add support for local cache path
+            const cachePath = this.host.globalTypingCachePath;
+            const tsdJsonPath = toPath("tsd.json", cachePath, getCanonicalFileName);
+            const tsdJson = this.host.getTsdJson(tsdJsonPath);
+            const { cachedTypingPaths, newTypingNames } = JsTyping.discoverTypings(
+                sys,
+                /* fileNames */ project.getFileNames(),
+                /* projectRootPath */ projectRootPath,
+                /* safeListPath */ undefined,
+                /* packageNameToTypingLocation */ tsdJson ? tsdJson.packageNameToTypingPath : undefined,
+                typingOptions,
+                compilerOptions
+            );
+
+            this.log("Cached typings include: " + cachedTypingPaths);
+            this.log("New typings to download: " + newTypingNames);
+
+            // Bail out when no actions are needed
+            if (cachedTypingPaths.length === 0 && newTypingNames.length === 0) {

```

Comment:
If either of this is not zero, something should be injected to the project, therefore the following process should not be skipped. Therefore it should be `&&` here